File: src/atcoder/newjudge2308/scala/newjudge2308_d.scala
import scala.io.StdIn
import scala.collection.mutable.ArrayBuffer
import scala.collection.mutable.Queue

def get2(): (Int, Int) =
  val line = StdIn.readLine().split(" ").map(_.toInt)
  return (line(0), line(1))

def get3(): (Int, Int, Int) =
  val line = StdIn.readLine().split(" ").map(_.toInt)
  return (line(0), line(1), line(2))

def depth(g: Array[ArrayBuffer[Int]], root: Int): Array[Int] =
  var depth = Array.fill(g.length) { -1 }
  depth(root) = 0
  var q = Queue[Int]()
  q += root
  while q.length > 0 do
    val v = q.dequeue
    for nv <- g(v) do
      if depth(nv) == -1 then
        depth(nv) = depth(v) + 1
        q += nv
  return depth

object Main extends App:
  var (n1, n2, m) = get3()
  var g = Array.fill(n1+n2) { ArrayBuffer[Int]() }
  for _ <- 0 until m do
    val (v, nv) = get2()
    g(v-1) += nv-1
    g(nv-1) += v-1
  
  val d1 = depth(g, 0)
  val d2 = depth(g, n1 + n2 - 1)
  val ans = d1.max + d2.max + 1
  println(ans)

File: src/atcoder/newjudge2308/ruby/newjudge2308_d.rb
n1, n2, m = gets.split.map(&:to_i)
g = Array.new(n1 + n2) { [] }
m.times do
  v, nv = gets.split.map(&:to_i)
  v -= 1
  nv -= 1
  g[v] << nv
  g[nv] << v
end

class Depth
  attr_reader :n, :g

  def initialize(g)
    @g = g
    @n = g.size
  end

  def solve(root)
    depth = Array.new(n, -1)
    depth[root] = 0
    q = [root]
    while q.size > 0
      v = q.shift
      g[v].each do |nv|
        next if depth[nv] != -1
        depth[nv] = depth[v] + 1
        q << nv
      end
    end
    depth
  end
end

d1 = Depth.new(g).solve(0)
d2 = Depth.new(g).solve(n1 + n2 - 1)
ans = d1.max + d2.max + 1
pp ans
